### ISSUE TITLE
Rust port

### ISSUE THREAD

**Issue by xnor104** [weetmuts/xmq#5]

**Comment by weetmuts** [comment]
Before doing more work on porting I had to resolve how to perform quoting properly. Check it out: https://weetmuts.github.io/xmq/

Please let me know what you think!

**Comment by xnor104** [comment]
Great job documenting the syntax, Frederik! I like that quoting is very intuitive and whitespace is respected/ignored as one would expect.

Some issues to clarify:
- You first introduced triple quotes for CDATA [here](https://github.com/weetmuts/xmq/blob/master/tests/test_005_cdata.sh). The new syntax now introduces raw string literals enabled the use of ' single quotes in strings. Does the new syntax now replace the CDATA syntax? I would prefer to drop the CDATA syntax in favor of raw string literals and instead make CDATA explicit somehow.
- We have to be careful when using examples for transforming XMQ to HTML (shown in one example). HTML allows tags without endtags (since it as based on SGML, not XML), i.e.: \<br>, \<input>. Also those tags don't require the short form syntax for closing, i.e.: \<br />; only \<br> required. We can either have an HTML mode for XMQ to respect such tags, or else we drop HTML examples. What do you think?

My status on the Rust port:
I havent worked on the port much yet. I played a bit writing down the grammar in [Pest](https://pest.rs/book/grammars/syntax.html).

**Comment by weetmuts** [comment]
Den mån 14 dec. 2020 kl 21:20 skrev xnor104 <notifications@github.com>:

> Great job documenting the syntax, Frederik! I like that quoting is very
> intuitive and whitespace is respected/ignored as one would expect.
>
Thanks!

> Some issues to clarify:
>
>    - You first introduced triple quotes for CDATA here
>    <https://github.com/weetmuts/xmq/blob/master/tests/test_005_cdata.sh>.
>    The new syntax now introduces raw string literals enabled the use of '
>    single quotes in strings. Does the new syntax now replace the CDATA syntax?
>    I would prefer to drop the CDATA syntax in favor of raw string literals and
>    instead make CDATA explicit somehow.
>
> You are quite right, I dropped the CDATA support, the spec/* files are the
foundation for the new tests,
the old tests are for the moment disabled. The new tests, test both the
binary supplied:
"./spec/genspechtml.sh xmq-binary" and they produce the xmq_spec.html which
is the one published
on weetmuts.github.io/xmq

I will clean up and re-active the old tests for more testing that is not
suitable to have in the spec.
There are some bugs in the incidental indentation removal code.

I still think cdata and pi nodes ought to be possible to express in xmq. So
how about:
    cdata{[ 'x y z' ]} generates <![CDATA[x y z]]>
    mypinode(? 'x y z' ?) generates <?mypinode x y z?>

This is possible since the current grammar forbids ( or { followed by a
character that is not a quote
and not a letter or underscore. But I would like to think a little more
before deciding on the syntax.
Cdata is really not recommended anymore, and pi nodes are rare, so we are
in no hurry. :-)

>
>    - We have to be careful when using examples for transforming XMQ to
>    HTML (shown in one example). HTML allows tags without endtags (since it as
>    based on SGML, not XML), i.e.: <br>, <input>. Also those tags don't require
>    the short form syntax for closing, i.e.: <br />; only <br> required. We can
>    either have an HTML mode for XMQ to respect such tags, or else we drop HTML
>    examples. What do you think?
>
> Actually,  I already added the html mode. :-) If it detects a <!DOCTYPE
html> or just <html>  then
it will set a flag in rapidxmxl to allow for void elements (br,img,hr etc)
I have already made a few modifications
to rapidxml to get the whitespace as I need, so adding the void elements
was pretty easy.
When converting from xmq to html, again it detects if first name is html,
then
it will write <!DOCTYPE html>. It will still write <br/> and <hr/> which is
permissible, but perhaps
we should reconsider that.

The reason I want to support html is because I realised that I will use xmq
for generating my html.
We can never expect to parse random slightly broken html on the net, but
for generating html and
examining the generated html, it should be fine.

>
>
> My status on the Rust port:
> I havent worked on the port much yet. I played a bit writing down the
> grammar in Pest <https://pest.rs/book/grammars/syntax.html>.
>
Neat! I guess the only part that will need some extra thinking is the rule:
use n+1 quotes to quote a string containing n quotes. It should be possible
to do in antlr, though...
E.g. use 5 sequential quotes to quote the string containing 4 sequential
quotes.
block='''''
 ''''a string '''containing a ''containing a 'string' '' ''' ''''
'''''

//Fredrik

**Comment by weetmuts** [comment]
>> I don't like having two syntaxes for the same output. The following are the same:
>> alfa { 'Hello World!' }
>> equals
>> alfa = 'Hello World!'
>> Though I understand that we need this syntax to represent inner text mixed with tags, like:
>> <p><em>Best</em> weather ever</p>

Yes, I agree, it is a quirk, but you could say that = is syntactic sugar.

I was thinking about forbidding the alfa{'x'} case and requiring
the user the write alfa=x. But then I thought about when you auto generate
a block of text, you execute a loop generating: list{'line1' 'line2' 'line3'}
but what if the loop will only run once? then you will get list{'line1'}
and that would be an error....and you have to add special code for the single line case.
But perhaps that is indeed acceptable. Hmmm.

Likewise I was toying with the idea of forcing the user to only use
the minimal number of single quotes to safely quote the content.

I have no firm opinion here, please let me know what you think.

**Comment by weetmuts** [comment]
>    Allowed identifiers should respect some rules to comply with XML standard, e.g.:
>        Identifier must start with either _ or a Unicode letter (latin a-z|A-Z, greek letter, etc.)
>        Can be separated once with :, then has another identifier starting with a name conforming to the first rule
>        Characters following the first letter can be numbers, letters, middledots and a few other bizarre Unicode characters

Thanks I will take a look.

**Comment by weetmuts** [comment]
> Generally I would like to be able to use XMQ as a data language like TOML, JSON, YAML and to skip the step to transform a document to XML. 

Yep, that is my intent as well. I intend to write a xmq parser, that can be combined with either a standalone xml api/library
or with a small xmq programming api. So you can choose to work with one of two different apis on the xmq data.
A lot of people who just want to read their config file, probably are hesitant to bring in an xml-library, but would
be happy to just use the xmq-parser-api. (Which is of course quite similar to the xml-api... ;-) )

> And/or I would like to be able to transform an XMQ document to HTML. 

Me too!

> This would require to omit the first output line <?xml version="1.0" encoding="UTF-8"?> and also to set the XMQ document into an HTML mode or a data language mode. But this is rather advanced and is out of scope for why XMQ was invented.

Already done!

**Comment by weetmuts** [comment]
Some updates pushed, now you can view your xmq file with colors and standardized format without converting to xml and back. xmq -v file.xmq

Been thinking about a=123 vs a{'123'}

I think both should be allowed, the underlying model is xml/html and everyone understands that, so
it should be easy to accept that a=123 is a syntactic sugar. Also a{'123'} is sometimes necessary when
constructing more complext strings. Like:
a{'' '   123' ''}
I.e. a="\n   123\n" in C-speak.

Also, I think it would be unnecessary annoying to limit the number of quotes to exactly the amount
needed to quote the underlying string.

**Comment by xnor104** [comment]
Again, great job and great ideas :)

- I agree with the declaration of `a = 123` being syntactic sugar for `a {'123'}`. Should be noted like that in the docs/specs.

- Im not sure about CDATA syntax. CDATA itself is so ugly that one should throw its ugliness right into the user's face. I would say the syntax should be as ugly as in XML: `node { <![CDATA[]]> }`. At least we shouldn't hide "CDATA", or the fact that special XML symbol like `< > &` are allowed inside a CDATA definition. Are XMQ special symbol allowed inside a CDATA declaration?

- The pi node syntax looks fine: `mypinode(? 'x y z' ?)`.

- To put XMQ into SGML/HTML mode with the specific doctype `<!DOCTYPE HTML>` it should be fine if the first node is `html { ... }`. But sometimes you want to use HTML snippets or use XMQ as a base for HTML templates or use a specific doctype. So you should be able to specifically put an XMQ document into SGML/HTML mode. Therefore we should allow the first line to be a `<!DOCTYPE>` declaration because the user himself may want to decide the parser mode/doctype. Or if the XMQ document starts with `= html`, it is meant to be an HTML snippet while the doctype is omitted. Are there any other void elements in different SGML dialects/doctypes? Maybe we should only support HTML then. I can't think of any other interesting use case.

- Like the latter point, XML versions should be declarable. I have no opinion on a syntax, though.

- Sequential quotes / raw string literals are easily parseable with the stack syntax of pest: https://pest.rs/book/grammars/syntax.html#the-stack-wip

**Comment by weetmuts** [comment]
Made some bug-fixes regarding incidental indentation removal and updated the spec with more thinking.

>Regarding CDATA: yeah, the correct syntax should be cdata {[ 'x y z' ]}. But I don't have a strong opinion on this. It's already so ugly in XML... It's discouraged anyway as you say.

I have for the moment decided to completely ignore CDATA, which means that any CDATA is converted into
the nominal text it represents i XMQ, then when XMQ is converted back to XML, it will use normal XML-escapes,
ie &amp; and &lt; and not CDATA. The XML-document should be equivalent. I think this is acceptable
since after loading an XML document we cannot see if it used a weird escape like &#65; for A.
I.e. we cannot anyway reproduce the exact same way the XML document quoted stuff, so lets just assume
CDATA is a weird quoting that we do not see. I assume some XML-parsers might expose the CDATA nodes,
some might just unquote the data and present it as text.

>Regarding SGML syntax <! ... >: ...Otherwise XMQ will become too complex....

Yes, I would like to stick to XML + minimal support necessary to generate/parse proper html5.

>But sometimes you want to use HTML snippets or use XMQ as a base for HTML templates or use a specific doctype. 

Yep, there is already a settings --html --nodec, which will produce/read snippets of html without the need for the doctype.

>To find an example for a syntax, how do we declare the doctype for
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.1//EN" "http://www.w3.org/TR/xhtml11/DTD/xhtml11.dtd"> in XMQ?

I do not know yet, I have to think about it.

> What about empty attributes in HTML? In HTML you can do stuff like this: <input required>.
 
Already supported. See the spec. :-)

> XML versions should be declarable in pi node syntax. The first line of the XMQ document could be: xml(? version = 1.0 encoding=UTF-8 standalone=yes ?)

Hmm, again I have to think about it. 

> Sequential quotes / raw string literals are easily parseable with the stack syntax of pest: https://pest.rs/book/grammars/syntax.html#the-stack-wip

Cool!

**Comment by xnor104** [comment]
You are working hard on xmq :) Im soon going to use it for creating a static site generator. Though Id like to make generated HTML code pass the [W3C validator](https://validator.w3.org/nu/).
This code is rejected due to self-closing `<head />`:
```
html(lang=de) {
	head{}
	body {
	br()
	img(src=abc.png alt=zapf)
	input(required)
	hr()}
}
```
(It's also rejected because `<head>` doesn't contain a title. If you add a title tag, the code passes the validator in all aspects).

Another bug I've encountered: if you add a `lang` attribute to the `html` tag, you get the XML header `<?xml version="1.0" encoding="UTF-8"?>` instead of a doctype. Valid HTML always requires setting a `lang` attribute.

**Comment by weetmuts** [comment]
Thanks! Now, when generating html, it only self-closes void tags. So <head/> and <p/>:
are now rendered as <head></head> and <p></p>. It cannot help with the missing title of course. :-)
The bug that prevented detection of html( is fixed and your test case is added to the test-suite.

I also refactored the code alot! So that I could build a library libxmq.so. 
Check the source to testur.cc and xmq.h.

As you can see in xmq_rapidxml.h there is a parser and render binding that works with rapidxml.
I intend to write a xmq_simplxml.h which is a very small xml implementation that you
can use if you only want to read an xmq config file and lookup the values.

**Comment by xnor104** [comment]
Haha nice you even used my code examples! xmq now works as expected :) Refactoring is very good since I can use a buffer now. Also the action hooks for rendering and parsing are very useful.

Im now looking into bridging C++ code to Rust.

Have an awesome new year, my friend!

**Comment by xnor104** [comment]
I've found a small bug: `p = Web Design` generates:
```
<p>Web</p>
<Design/>
```
Quoting `'Web Design'` solves the issue.

**Comment by weetmuts** [comment]
Its not a bug. :-) Whitespace (space,tab,cr,lf) must always be quoted if its not an xmq delimiter.

**Comment by weetmuts** [comment]
My intention when generating html is to be able to generate multi-language pages, ie you store the text to be inserted somehwere else and only store language markers:s to be replaced with longer text. So, I will write:
h1=intro
div(id=123) { 
  p=explain_web_design
  p=colors_and_css
}

Then I have a tool that replaces:
intro with 'Introduction to HTML design'
explain_web_design with 'The history of web design is long and convoluted, it all began in 1897 when....'
colors_and_css with 'Colors are best configured in the css. Pick for example...'
etc

The replacement can of course be done on the xmq file itself, but perhaps better is to
make your own subclass of RenderActionsRapidXML or write your own from scratch.
That checks in appendData(void *parent, Token t), if the token matches something that
should be replaced and thus inserts that into the html instead.

**Comment by xnor104** [comment]
Re not-a-bug, I see :)
Im now building my [personal website](https://github.com/xnor104/fey-website) using xmq!
Re xmq templating. Good idea. I will be using xmq for describing multilingual forms inside a single xmq file. One version of a form will be used per translated page. I will let you know when I advance.

**Comment by weetmuts** [comment]
Cool!

**Comment by weetmuts** [comment]
I am generating my html now, and I found out that newlines should not be inserted before/after inline elements.
I have pushed a fix for this.

In particular this was a problem:

a(href=goto.html) { img(src=img.png) }

This would generate:

<a href="goto.html">
  <img src="img.png"/>
</a>
Which means the browser will add a space before and after the image. Not good.
An important aspect of xmq is that all significant whitespace should be visible inside quotes.

Now (if using --html or the first tag is html) it generates:
<a href="goto.html"><img src="img.png"/></a>

It uses the list of default elements to detect inline elements. 

Unfortunately since the CSS can change this, for those
cases you have to run with a future option --noindent that will
prevent all prettyprinting of the html/xml. Or we have to 
tell xmq: --html --inlinelements="a img"

I would like to try to keep some sort of pretty printing
of the html possible. So for the moment, xmq tries
to know when pretty printing is permissable.
We will see if this actually is workable in the future.

**Comment by xnor104** [comment]
Yes, good point. I encountered the same problem with lists and whitespace within \<li>\</li> and \<ul>\</ul>. Using CSS I circumvented the problem by setting `ul { font-size: 0}` and `li { font-size: 1rem }`.

**Comment by weetmuts** [comment]
@xnor104 I made a tiny change to the grammar in a corner case.
Now two consecutive quotes are not joined using a newline, they are joined seamlessly.
Probably does not affect you, unless you work with explicit indenting in xml/html.

**Comment by weetmuts** [comment]
@xnor104 after a complete rewrite, xmq is now based in libxml2 so we get all the xslt/xsd stuff as well.
Checkout https://libxmq.org

And I hope you have a Happy New Year!
